File: deploy/identity-webhook/billing-identity.mjs
import { createHash } from "node:crypto";

export const CUSTOMER_KEY_PREFIX = "ch_";

const BASE32_ALPHABET = "abcdefghijklmnopqrstuvwxyz234567";

function encodeBase32LowerNoPadding(buffer) {
  let bits = 0;
  let value = 0;
  let output = "";
  for (const byte of buffer) {
    value = (value << 8) | byte;
    bits += 8;
    while (bits >= 5) {
      output += BASE32_ALPHABET[(value >>> (bits - 5)) & 31];
      bits -= 5;
    }
  }
  if (bits > 0) {
    output += BASE32_ALPHABET[(value << (5 - bits)) & 31];
  }
  return output;
}

export function buildAuthId(tenantId, clientId, externalUserId) {
  return `${String(tenantId).trim()}:${String(clientId).trim()}:${String(externalUserId).trim()}`;
}

export function parseAuthId(authId) {
  const trimmed = String(authId || "").trim();
  const first = trimmed.indexOf(":");
  if (first <= 0 || first >= trimmed.length - 1) {
    return null;
  }
  const secondOffset = trimmed.slice(first + 1).indexOf(":");
  if (secondOffset <= 0) {
    return null;
  }
  const second = first + 1 + secondOffset;
  if (second >= trimmed.length - 1) {
    return null;
  }
  return {
    tenantId: trimmed.slice(0, first),
    clientId: trimmed.slice(first + 1, second),
    externalUserId: trimmed.slice(second + 1),
  };
}

export function parseLegacyAuthId(authId) {
  const trimmed = String(authId || "").trim();
  const first = trimmed.indexOf(":");
  if (first <= 0 || first >= trimmed.length - 1) {
    return null;
  }
  return {
    clientId: trimmed.slice(0, first),
    externalUserId: trimmed.slice(first + 1),
  };
}

export function buildCustomerKey(tenantId, clientId, externalUserId) {
  const authId = buildAuthId(tenantId, clientId, externalUserId);
  const digest = createHash("sha256").update(authId).digest();
  return `${CUSTOMER_KEY_PREFIX}${encodeBase32LowerNoPadding(digest)}`;
}

File: tenant-admin/internal/billingidentity/billingidentity.go
package billingidentity

import (
	"crypto/sha256"
	"encoding/base32"
	"fmt"
	"strings"
)

const (
	CustomerKeyPrefix = "ch_"
	MaxSegmentLength  = 120
)

var base32NoPadding = base32.StdEncoding.WithPadding(base32.NoPadding)

func ValidateSegment(name string, value string) error {
	trimmed := strings.TrimSpace(value)
	if trimmed == "" {
		return fmt.Errorf("%s must be non-empty", name)
	}
	if len(trimmed) > MaxSegmentLength {
		return fmt.Errorf("%s must be <= %d chars", name, MaxSegmentLength)
	}
	if strings.Contains(trimmed, ":") {
		return fmt.Errorf("%s cannot contain ':'", name)
	}
	return nil
}

func BuildAuthID(tenantID string, clientID string, externalUserID string) (string, error) {
	if err := ValidateSegment("tenantId", tenantID); err != nil {
		return "", err
	}
	if err := ValidateSegment("clientId", clientID); err != nil {
		return "", err
	}
	if err := ValidateSegment("externalUserId", externalUserID); err != nil {
		return "", err
	}
	return strings.TrimSpace(tenantID) + ":" + strings.TrimSpace(clientID) + ":" + strings.TrimSpace(externalUserID), nil
}

func ParseAuthID(authID string) (string, string, string, bool) {
	trimmed := strings.TrimSpace(authID)
	if trimmed == "" {
		return "", "", "", false
	}
	first := strings.Index(trimmed, ":")
	if first <= 0 || first >= len(trimmed)-1 {
		return "", "", "", false
	}
	secondOffset := strings.Index(trimmed[first+1:], ":")
	if secondOffset <= 0 {
		return "", "", "", false
	}
	second := first + 1 + secondOffset
	if second >= len(trimmed)-1 {
		return "", "", "", false
	}
	return trimmed[:first], trimmed[first+1 : second], trimmed[second+1:], true
}

func ParseLegacyAuthID(authID string) (string, string, bool) {
	trimmed := strings.TrimSpace(authID)
	if trimmed == "" {
		return "", "", false
	}
	colon := strings.Index(trimmed, ":")
	if colon <= 0 || colon >= len(trimmed)-1 {
		return "", "", false
	}
	return trimmed[:colon], trimmed[colon+1:], true
}

func BuildCustomerKey(tenantID string, clientID string, externalUserID string) (string, error) {
	authID, err := BuildAuthID(tenantID, clientID, externalUserID)
	if err != nil {
		return "", err
	}
	sum := sha256.Sum256([]byte(authID))
	return CustomerKeyPrefix + strings.ToLower(base32NoPadding.EncodeToString(sum[:])), nil
}
